yaptide/persistence/models.py

```diff
@@ -209,7 +209,9 @@ def update_state(self, update_dict: dict):  # skipcq: PY-R1000
             self.requested_primaries = update_dict["requested_primaries"]
         if "simulated_primaries" in update_dict and self.simulated_primaries != update_dict["simulated_primaries"]:
             self.simulated_primaries = update_dict["simulated_primaries"]
-        if "task_state" in update_dict and self.task_state != update_dict["task_state"]:
+        if "task_state" in update_dict and self.task_state != update_dict["task_state"] and not (
+                self.task_state in [EntityState.COMPLETED.value, EntityState.FAILED.value]
+                and update_dict["task_state"] == EntityState.CANCELED.value):
```

Comment:
the logic here is too complex, readability of this line sucks.
Can you break it down into several nested lines ?
Some comments may be necessary, for example

```suggestion
        if "task_state" in update_dict:
             # check if update is needed
        else:
           # do nothing, as nothing related to task_update came with update dict
            pass
```